### PR TITLE
MCO-113: Preserve previous boot MCD log

### PR DESCRIPTION
**Pull request by djoshy**
**- What I did**
The daemon will begin to save its logs right before shutdown at `/etc/machine-config-daemon/previous-logs/`. I'm hoping that we can update the must gather tool so that it is able to scoop up this log file. I also wanted this to be a backportable solution so I tried to keep it out of the critical path, on encountering an error the daemon logs it, but it's not fatal.

**- How to verify it**
- Roll out an MC change that causes a node reboot. A cluster upgrade should also trigger this. 
- On reboot, debug into the node, and you should see the previous daemon logs saved at `/etc/machine-config-daemon`:
```
sh-5.1# ls /etc/machine-config-daemon/previous-logs/openshift-machine-config-operator_machine-config-daemon-qmgx4_a2b53ce6-ac08-4211-b93f-c1286d4e0554/machine-config-daemon/
0.log  1.log
```
If you're lucky, you might also the see the daemon logs indicate successful save. Since this happens late in the daemon shutdown process, the line does get lost occasionally:
```
2024-12-04T18:55:32.588709698+00:00 stderr F I1204 18:55:32.588667   12126 daemon.go:2930] Daemon logs from /var/log/pods/openshift-machine-config-operator_machine-config-daemon-qmgx4_a2b53ce6-ac08-4211-b93f-c1286d4e0554 preserved at /etc/machine-config-daemon/previous-logs/openshift-machine-config-operator_machine-config-daemon-qmgx4_a2b53ce6-ac08-4211-b93f-c1286d4e0554
2024-12-04T18:55:32.588804462+00:00 stderr F I1204 18:55:32.588787   12126 daemon.go:1400] Shutting down MachineConfigDaemon

```

### CODE REVIEW

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -2889,3 +2890,43 @@ func (dn *Daemon) getUnsupportedPackages() {
...
+		}
+		return nil
+	}); walkErr != nil {
+		klog.Errorf("Daemon logs could could not be found due to error: %v", walkErr)
+		return
+	}
+
+	baseDir := "/etc/machine-config-daemon/previous-logs/"
+	// Remove the old logs if it exists
+	if err := os.RemoveAll(baseDir); err != nil {
```
Comment (djoshy) [inline pkg/daemon/daemon.go:2913]: I opted to remove any old logs as I was a bit worried that we might fill up the disk with every daemon restart. I also figured the (n-1) set of logs are most critical, but I'm open to other approaches here too :thinking:

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -2889,3 +2890,43 @@ func (dn *Daemon) getUnsupportedPackages() {
...
+
+	// Ensure the base directory exists
+	if err := os.MkdirAll(baseDir, 0755); err != nil {
+		klog.Errorf("Failed to create logs directory: %v", err)
+		return
+	}
+
+	// Create preservation location; this will change based on the daemon name/uid
+	preservationLocation := baseDir + logDir
+	cmd := exec.Command("cp", "--recursive", currentLogLocation, preservationLocation)
```
Comment (djoshy) [inline pkg/daemon/daemon.go:2926]: Easier to do a recursive file copy via shell command instead of walking the directory, so I went this route.

**Review by yuqi-zhang — APPROVED** [review]
/lgtm
Should be a safe addition. I guess since it's a shutdown func, it's possible for a large enough log dir that the copy doesn't finish?

**Comment by djoshy** [comment]
> /lgtm Should be a safe addition. I guess since it's a shutdown func, it's possible for a large enough log dir that the copy doesn't finish?

Hmm, wouldn't this actually block on until the copy is complete? It is a defer call on the `daemon.Run()`. If that is the case, perhaps it wouldn't hurt to wrap the copy in a timeout?

/hold

**Pull request merged**